### ISSUE TITLE
EF Core 7 support

### ISSUE THREAD

**Issue by lauxjpn** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1557]
We will use this issue to track the progress on supporting EF Core 7, including its previews.

**Comment by JustinErdmier** [comment]
With EF Core 7 making headway, are there any new features that this library hasn't implemented yet or should we be okay to start using EF Core 7?

**Comment by glen-84** [comment]
I'm getting:

> Method not found: 'Void CoreTypeMappingParameters..ctor(System.Type, Microsoft.EntityFrameworkCore.Storage.ValueConversion.ValueConverter, Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer, Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer, System.Func`3<Microsoft.EntityFrameworkCore.Metadata.IProperty,Microsoft.EntityFrameworkCore.Metadata.IEntityType,Microsoft.EntityFrameworkCore.ValueGeneration.ValueGenerator>)'.

> at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlBoolTypeMapping..ctor(String storeType, Nullable``1 dbType, Nullable`1 size)\n   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlTypeMappingSource..ctor(TypeMappingSourceDependencies dependencies, RelationalTypeMappingSourceDependencies relationalDependencies, IMySqlOptions options)\n

Any news on the support for EF 7?

**Comment by AlexCastroAlex** [comment]
> I'm getting:
> 
> > Method not found: 'Void CoreTypeMappingParameters..ctor(System.Type, Microsoft.EntityFrameworkCore.Storage.ValueConversion.ValueConverter, Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer, Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer, System.Func`3<Microsoft.EntityFrameworkCore.Metadata.IProperty,Microsoft.EntityFrameworkCore.Metadata.IEntityType,Microsoft.EntityFrameworkCore.ValueGeneration.ValueGenerator>)'.
> 
> > at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlBoolTypeMapping..ctor(String storeType, Nullable``1 dbType, Nullable`1 size)\n   at Pomelo.EntityFrameworkCore.MySql.Storage.Internal.MySqlTypeMappingSource..ctor(TypeMappingSourceDependencies dependencies, RelationalTypeMappingSourceDependencies relationalDependencies, IMySqlOptions options)\n
> 
> Any news on the support for EF 7?

Hello all , 

I'm having the same issue on EF Core 6.0 on project build on SDK 7.0. 
Is it related to .NET 7 or am i missing something ?

Thanks in advance !
Regards

**Comment by bgrainger** [comment]
FYI, [MySqlConnector 2.2.0 Beta 1](https://www.nuget.org/packages/MySqlConnector/2.2.0-beta.1) is now available with .NET 7.0 support. (This shouldn't be _required_ for EF Core 7.0--I think v2.1.13 should still work--but could be the new minimum dependency for Pomelo v7.x.)

**Comment by lauxjpn** [comment]
The current plan is to have basic EF Core 7 support out on Monday.

**Comment by meriturva** [comment]
Just to know, what about JSON Columns support? will be included?

**Comment by mguinness** [comment]
Pomelo provider already has support for JSON.  See similar issue with Npgsql provider at https://github.com/npgsql/efcore.pg/issues/2452.

**Comment by AlexCastroAlex** [comment]
Hi there ! any news on EF 7  support as it was supposed to be on monday  ? 

Reagrds

**Comment by lauxjpn** [comment]
@AlexCastroAlex It is going to take us a couple of days longer. There will be a PR on Wednesday, to track the progress. Something testable should be available at the end of the week.

**Comment by AlexCastroAlex** [comment]
> @AlexCastroAlex It is going to take us a couple of days longer. There will be a PR on Wednesday, to track the progress. Something testable should be available at the end of the week.

Thanks for answer ! strength and honor ! 

Regards

**Comment by lauxjpn** [comment]
The PR to track the progress is #1717.
I expect the remaining test to have become green in the next few days.

**Comment by ekjuanrejon** [comment]
@lauxjpn any updates for this week?

**Comment by VladislavAntonyuk** [comment]
Tomorrow we'll see .NET 7. Do we need rc2 release?

**Comment by solrevdev** [comment]
Hi, 

Just upgrading to .NET 7  / EF Core 7 and it seems the latest version of Pomelo.EntityFrameworkCore.MySql version 6.0.2 does not support it

```xml
<PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="6.0.2" />
```

I get the following as my only error: 

Build FAILED.

error NU1608: Detected package version outside of dependency constraint: Pomelo.EntityFrameworkCore.MySql 6.0.2 requires Microsoft.EntityFrameworkCore.Relational (>= 6.0.7 && < 7.0.0) but version Microsoft.EntityFrameworkCore.Relational 7.0.0 was resolved.

**Comment by zivfuture** [comment]
> Hi,
> 
> Just upgrading to .NET 7 / EF Core 7 and it seems the latest version of Pomelo.EntityFrameworkCore.MySql version 6.0.2 does not support it
> 
> ```
> <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="6.0.2" />
> ```
> 
> I get the following as my only error:
> 
> Build FAILED.
> 
> error NU1608: Detected package version outside of dependency constraint: Pomelo.EntityFrameworkCore.MySql 6.0.2 requires Microsoft.EntityFrameworkCore.Relational (>= 6.0.7 && < 7.0.0) but version Microsoft.EntityFrameworkCore.Relational 7.0.0 was resolved.

same issue as you

**Comment by aL3891** [comment]
Blocked from upgrading due to this as well, i know oss development is often a thankless job so let me start off by saying: Thanks for all the work you do!
However if you have any timeline even if its a ways of it would be very helpful :)

**Comment by zijianhuang** [comment]
### Steps to reproduce
1. Update existing .NET 7 application to .NET 7. No problem.
2. Update packages to EntityhFrameworkCore 7, error:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	NU1107	Version conflict detected for Microsoft.EntityFrameworkCore.Relational. Install/reference Microsoft.EntityFrameworkCore.Relational 7.0.0 directly to project PoemsMyDbCreator to resolve this issue. 
 PoemsMyDbCreator -> Fonlow.PoemsApp.EF -> Microsoft.EntityFrameworkCore.Relational (>= 7.0.0) 
 PoemsMyDbCreator -> Pomelo.EntityFrameworkCore.MySql 6.0.2 -> Microsoft.EntityFrameworkCore.Relational (>= 6.0.7 && < 7.0.0).	
```


### Further technical details

MySQL version: 8
Operating system: Windows 10
Pomelo.EntityFrameworkCore.MySql version: 6.0.2
Microsoft.AspNetCore.App version: 7

**Comment by glen-84** [comment]
Please stop spamming this issue. There is a work-in-progress PR (#1717).

**Comment by mapo80** [comment]
I'm using [7.0.0-alpha.1](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/7.0.0-alpha.1) version.

I'll give you feedback about this version.

Great work!!

**Comment by lauxjpn** [comment]
Prerelease [7.0.0-alpha.1](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/7.0.0-alpha.1) (compatible with EF Core 7.0) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/7.0.0-alpha.1).

This includes full support of all features of EF Core 7.
We also completed migration and scaffolding support for MariaDB sequences.

The current plan is to release a `7.0.0-beta.1` version in a week and a stable official version by the end of the month.

**Comment by areltasarim** [comment]
Database is created but tables are not created

**Comment by alexandrecastro-bib** [comment]
> Database is created but tables are not created

Check if all your tables have Primary key as i had an issue related to EF Core 7.0 forcing to have Primary Keys

**Comment by zijianhuang** [comment]
I have 2 complex business applications each with fairly complex databases and comprehensive integration test suites, all work well 7.0.0-alpha.1. As always, even with ORM, having the code first model design conforming to database normalization is good for many things. 
I was not aware of EF Core 7.0 forcing to have Primary Keys, however, I do think this is a good move.

**Comment by areltasarim** [comment]
all my id fields are like this anyway

[Key, DatabaseGenerated(DatabaseGeneratedOption.Identity)]

**Comment by mguinness** [comment]
@areltasarim Create a new issue w/ an example project so it can be looked into.

**Comment by areltasarim** [comment]
You can download the project from the link below

https://github.com/areltasarim/Pomelo

the default project as in the picture below EticaretWebCoreMigrationMysql Let's make an update database by selecting

![Başlıksız-1](https://user-images.githubusercontent.com/101069343/202844574-74c6181b-3f41-4e38-812c-0ccce37ce36f.png)

**Comment by alexandrecastro-bib** [comment]
> You can download the project from the link below
> 
> https://github.com/areltasarim/Pomelo
> 
> the default project as in the picture below EticaretWebCoreMigrationMysql Let's make an update database by selecting
> 
> ![Başlıksız-1](https://user-images.githubusercontent.com/101069343/202844574-74c6181b-3f41-4e38-812c-0ccce37ce36f.png)

I can't access your project but did you generate your first migration ? 
If not, this is why you don't have tables.

**Comment by areltasarim** [comment]
you can access the project now

https://github.com/areltasarim/Pomelo

**Comment by areltasarim** [comment]
normally I did add-migration in my project, but I forgot this in the project I sent you, you add-migration first

**Comment by areltasarim** [comment]
re-updated the project can you download again

**Comment by lauxjpn** [comment]
@areltasarim I moved your conversation to the new issue #1726.

Please keep this issue here clean, since it is being tracked by other members of the community as well.
Always open a new issue, if you encounter specific problems.
Thank you.

**Comment by malthuis** [comment]
@lauxjpn could you give an update on when you think the beta or stable version are going to be released?

Thank you for your efforts :)

**Comment by kampilan** [comment]
Why are comments asking for an update on the plan for stable release being marked as off topic? Isn't it easier to simply give an update? No news is always inferior to bad news.

**Comment by malthuis** [comment]
> Why are comments asking for an update on the plan for stable release being marked as off topic? Isn't it easier to simply give an update? No news is always inferior to bad news.

especially since there were concrete dates when to look again (in a week, by the end of the month). Letting them pass without any communication does not make much sense.

**Comment by mguinness** [comment]
Work on [7.0.0-alpha.2](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/46) is ongoing.  I assume that if no new bugs are reported then a stable release will be the next one.  I don't have any information on the timeline.  It should be noted that the Oracle provider is still in preview too with a lot more developer resources available.  Just a reminder that we welcome additional contributors as already mentioned in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1717#issuecomment-1311986697 and that new issue https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1731 requires investigation if you are able to help.

**Comment by jscottPTC** [comment]
@lauxjpn thanks for working on this issue.  I did not see any way to donate to you or this project.  Do you have a paypal or other way to receive compensation?

**Comment by kampilan** [comment]
> @lauxjpn thanks for working on this issue. I did not see any way to donate to you or this project. Do you have a paypal or other way to receive compensation?

Yes! Find a way to monetize this project. You and the project deserve it. From a selfish perspective, I feel more comfortable using projects that make money. Take a look at Fody and the way they do it. You could do the same thing. Better then nothing.

**Comment by renemadsen** [comment]
@mguinness do we have any progress on the release?
We are using the alpha in testing env without issues. This includes a lot of github repositories with automated tests. So I would say the alpha is ready to go into a stable release. Also all tests are passing, so what is holding us back from a release?

I'm happy to help out with this project with test automation and the daily maintenance to make sure we are up to date, since this project is a key component of our own projects.

**Comment by mguinness** [comment]
The main developer hasn't been active on GitHub this month, meaning they're either busy at work or taking a break.  With the holidays coming up I wouldn't anticipate a new release until January now.  As you already noted there shouldn't be much additional work required, so I suspect we'll have something by the end of January and have a stable release out before the Oracle provider.  We also need to revisit [joining the .NET Foundation](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/646) and see how we can best use your help that you're offering.

**Comment by lauxjpn** [comment]
Okay, so it looks like there have been no general issues with the `7.0.0-alpha.1` release.

This is our version of an RC release when it is not correlated to an EF Core RC release, so it expresses being stable and feature complete.

There are no changes in regards to `7.0.0-alpha.1`, except making sure the correct exception is being thrown when using an incorrect connection string in an already opened connection (see #1719).

**Comment by renemadsen** [comment]
> With the holidays coming up I wouldn't anticipate a new release until January now. As you already noted there shouldn't be much additional work required, so I suspect we'll have something by the end of January and have a stable release out before the Oracle provider. We also need to revisit [joining the .NET Foundation](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/646) and see how we can best use your help that you're offering.

Hope we can move to a more inline release cycle with EF Core and as I stated, I can gladly help with that process.

**Comment by renemadsen** [comment]
> Prerelease [7.0.0-silver.1](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/7.0.0-silver.1) (compatible with EF Core 7.0.x) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/7.0.0-silver.1).
> 
> Unless we need to publish another _silver_ version, the next release will be `7.0.0` (GA).

I know I'm pushing it @lauxjpn , but what is the reason for alpha and now silver, when all tests are passing?

**Comment by lauxjpn** [comment]
> I know I'm pushing it @lauxjpn , but what is the reason for alpha and now silver, when all tests are passing?

Having all 26K tests succeeding is actually a requirement for all our commits, so therefore _all_ our releases are generally of very high quality (which also includes our nightly builds).

While the test coverage is huge, there are parts of the provider, that are covered better than others. For example, there are too many possible combinations of operations and database versions to test all of them, especially when it comes to migrations.

Our common release strategy is the incrementally publish prereleases with an increasingly higher commitment from us in regards to stability (see [Version.props](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/e8377890f1baffbb52eebbce9f5f14687789a83d/Version.props#L4-L13)), so that more members of the community test their real world apps before we go GA. While we could of course just release an official version right away, any update related bug that we missed would then be encountered by many more people in production systems, which is costly.

So therefore we would always release at least an RC/silver version, so that everybody can feel encouraged to test their apps with code that we consider finished, but before we have made it an official release. Since testing a real world system takes some time, we should give it at least a week, before we go ahead and release a GA version.

@renemadsen If you want to use EF Core 7 with Pomelo then please feel encouraged do so! I have been using Pomelo 7 in most of my projects since `7.0.0-alpha.1`.

(When it comes to critical applications though, I personally never jump on the latest freshly released technology or major version, but always wait at least a month or two. Saves time and sanity.)

**Comment by codelovercc** [comment]
Hi, every one! I upgrade Microsoft.EntityFrameworkCore to 7.0.1 and I get a Warn message.

Detected package version outside of dependency constraint: Pomelo.EntityFrameworkCore.MySql 6.0.2 requires Microsoft.EntityFrameworkCore.Relational (>= 6.0.7 && < 7.0.0) but version Microsoft.EntityFrameworkCore.Relational 7.0.1 was resolved.

**Comment by mguinness** [comment]
@codelovercc To install `7.0.0-silver.1` please read [Installing and updating pre-release packages](https://learn.microsoft.com/en-us/nuget/create-packages/prerelease-packages#installing-and-updating-pre-release-packages).

**Comment by nightcoder62** [comment]
I want to mention this here, because it's probably very much related to EF7: #1740

**Comment by sander1095** [comment]
Hey @lauxjpn, not to be greedy (and I hate to be that guy) but I am wondering if you have another ETA on the 7.x.x release? 

Is it still around the end of january? Looking at the 7.x.x. milestone, there seems to be a lot of work needing to get done. https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/milestone/39

Looking forward to your reply!

**Comment by lauxjpn** [comment]
Release [7.0.0](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/7.0.0) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/7.0.0).

The only change/fix since `7.0.0-silver.1` is #1744.

Thanks to everybody for being patient and especially to everybody who contributed!
